ibis/common/scope.py

```diff
@@ -0,0 +1,149 @@
+    {
+        op[ibis.expr.operations.Node]: {
+            "value": v[Object]
+            "timecontext": t[TimeContext]
+        }
+        ...
+    }
+
+    `scope` uses `op` as the key, `v` is the cached result of executing `op`,
+    the type of the result may differ in different backends. `t` is the time
+    context associate with the result, set it to None if there is no time
+    context in execution.
+
+    When there are no time contexts associate with the cached result, getting
+    and setting values in Scope would be as simple as get and set in a normal
+    dictonary. With time contexts, we need the following logic for getting
+    and setting items in scope:
+
+    Set scope kv pair: before setting the value op in scope we need to perform
+    the following check first:
+
+    Test if op is in scope yet
+    - No, then put op in scope, set timecontext to be the current timecontext
+    (None if timecontext is not present), set value to be the DataFrame or
+    Series of the actual data.
+    - Yes, then get the timecontext stored in scope for op as old_timecontext,
+    and compare it with current timecontext:
+    If current timecontext is a subset of old timecontext, that means we
+    already cached a larger range of data. Do nothing and we will trim data in
+    later execution process.
+    If current timecontext is a superset of old timecontext, that means we
+    need to update cache. Set value to be the current data and set timecontext
+    to be the current timecontext for op.
+    If current timecontext is neither a subset nor a superset of old
+    timcontext, but they overlap, or not overlap at all. For example this will
+    happen when there is a window that looks forward, over a window that looks
+    back. So in this case, we should not trust the data stored either, and go
+    on to execute this node. For simplicity, we update cache in this case as
+    well.
+
+"""
+from typing import Optional
+
+from ibis.expr.timecontext import TimeContextRelation, compare_timecontext
+from ibis.expr.typing import TimeContext
+
+
+class Scope:
+    def __init__(self, items):
+        self.scope = items
+
+    def get_scope(self):
+        """ Get all items in scope.
+        """
+        return self.scope
+
+    @staticmethod
+    def make_scope(op, result, timecontext: Optional[TimeContext]):
+        """make a Scope class, adding (op, result, timecontext) into the
+           scope
+
+        Parameters
+        ----------
+        op: ibis.expr.operations.Node, key in scope.
+        result : scalar, pd.Series, pd.DataFrame, concrete data.
+        timecontext: Optional[TimeContext], time context associate with the
+        result.
+
+        Returns
+        -------
+        Scope: a new Scope class with op in it.
+        """
+        return Scope({op: {'value': result, 'timecontext': timecontext}})
+
+    @staticmethod
+    def from_scope(other_scope):
+        """make a Scope class, copying other_scope
```

Comment:
This can just be the standard copy method.

https://stackoverflow.com/questions/1500718/how-to-override-the-copy-deepcopy-operations-for-a-python-object

Reply:
I just change it to copy, moreover, I find it will break to use deepcopy, which will cause an error in aggregation in testing `frame[name] is obj`